### PR TITLE
ISSUE-28: bring key rotation changes

### PR DESCRIPTION
**Pull request by huyumars**

### CODE REVIEW

File: src/infra/util/RaftClusterResolver.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/************************************************************************
```
Comment (jackyjia) [inline src/infra/util/RaftClusterResolver.h:1]: - [ ] this file is not needed

File: src/infra/util/Util.h [hunk 2]
```diff
@@ -109,8 +109,28 @@ class Util final {
     return data;
   }
 
+  /**
```
Comment (jackyjia) [inline src/infra/util/Util.h:112]: - [ ] below method is not needed

File: test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key [hunk 3]
```diff
@@ -0,0 +1,5 @@
+keys {
```
Comment (jackyjia) [inline test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key:1]: - [ ] revert this change as it's not needed

File: test/infra/raft/cluster_config/three_versions.pu_log_encrypt_symmetric.key [hunk 4]
```diff
@@ -0,0 +1,13 @@
+keys {
```
Comment (jackyjia) [inline test/infra/raft/cluster_config/three_versions.pu_log_encrypt_symmetric.key:1]: ditto

File: test/infra/raft/cluster_config/two_versions.pu_log_encrypt_symmetric.key [hunk 5]
```diff
@@ -0,0 +1,9 @@
+keys {
```
Comment (jackyjia) [inline test/infra/raft/cluster_config/two_versions.pu_log_encrypt_symmetric.key:1]: ditto

File: test/infra/util/UtilTest.cpp [hunk 6]
```diff
@@ -18,12 +18,31 @@ limitations under the License.
 
 namespace gringofts::test {
 
-TEST(UtilTest, getCurrentVersionTest) {
+TEST(UtilTest, ReleaseVersionTest) {
```
Comment (jackyjia) [inline test/infra/util/UtilTest.cpp:21]: - [ ] do not keep this method

File: test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key (hunk 3 above)
Comment (jackyjia) [inline test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key:1]: - [ ] is this a pesudo key?

File: test/infra/raft/cluster_config/three_versions.pu_log_encrypt_symmetric.key [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  version: 3
+  key: "InFQSEfe7YuDvKsGX8GeYfQQOgRGjeauI4yc5qQMph4="
+}
```
Comment (jackyjia) [inline test/infra/raft/cluster_config/three_versions.pu_log_encrypt_symmetric.key:4]: ditto

File: test/infra/raft/cluster_config/two_versions.pu_log_encrypt_symmetric.key (hunk 5 above)
Comment (jackyjia) [inline test/infra/raft/cluster_config/two_versions.pu_log_encrypt_symmetric.key:1]: ditto

File: test/infra/util/config/aes.key [hunk 8]
```diff
@@ -0,0 +1,13 @@
+keys {
```
Comment (jackyjia) [inline test/infra/util/config/aes.key:1]: pesudo key?

File: test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key (hunk 3 above)
Comment (huyumars) [reply test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key:1]: yes

**Comment by jackyjia** [comment]
- [ ] append issue # to each commit message

File: src/infra/util/StrUtil.h [hunk 9]
```diff
@@ -43,6 +43,33 @@ class StrUtil final {
     return buffer;
   }
 
+  /**
+   * Compliance with Golang multi-lines
```
Comment (qiawu) [inline src/infra/util/StrUtil.h:47]: A minor comment: these three functions are not used in Gringofts

File: test/infra/raft/cluster_config/fixed_member_raft_node3.ini [hunk 10]
```diff
@@ -17,3 +17,7 @@ segment.meta.size.limit = 4194304      ; 4MB
 [streaming]
 grpc.port               = 5678
 max.concurrency         = 4
+
+[aes]
+enable     = true          ; enable HMAC
+filename   = ../test/infra/raft/cluster_config/three_versions.pu_log_encrypt_symmetric.key
```
Comment (qiawu) [inline test/infra/raft/cluster_config/fixed_member_raft_node3.ini:23]: I guess we may need to rename this file

File: test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key (hunk 3 above)
Comment (qiawu) [reply test/infra/raft/cluster_config/one_version.pu_log_encrypt_symmetric.key:1]: I feel it is not appropriate to include these encryption related test into Gringofts

**Comment by huyumars** [comment]
retest

**Comment by jackyjia** [comment]
@qiawu could you review the PR again?

**Review by qiawu — APPROVED** [review]
LGTM. Thx !

**Review by qiawu — APPROVED** [review]
LGTM. Thx !

**Review by qiawu — DISMISSED** [review]

File: test/infra/raft/v2/FixedMembershipTest.cpp [hunk 11]
```diff
@@ -71,15 +78,131 @@ TEST_F(FixedMembershipTest, NormalTest) {
...
   mClusterUtil.killAllServers();
+  mClusterUtil.disableAllSyncPoints();
 }
 
+/// in this set up, server1 has 1 version, server2 has 2 versions, server3 has 3 versions
+/// if we elect server1, then the cluster works fine
+/// if we elect server2, then server1 will crash
+/// if we elect server3, then server1/2 will crash
+/// TODO qiawu will try to fix this test
+//  TEST_F(FixedMembershipTest, SecretKeyVersionMissMatchTest) {
```
Comment (jackyjia) [inline test/infra/raft/v2/FixedMembershipTest.cpp:98]: - [ ] disable the test instead of commenting it out.

**Review by jackyjia — APPROVED** [review]
LGTM. Thanks!
